### PR TITLE
Extend `branches_sharing_code` to match arms with a shared tail

### PR DESCRIPTION
**Pull request by sylvestre**
This extends `branches_sharing_code` to also fire on `match` expressions, not just `if`/`else`.

### What it does

When every arm of a `match` is a block ending in the **same trailing expression**, and the `match` is itself in tail position, that expression can be hoisted out below the `match`:

```rust
match mode {
    Mode::A => { a(); Ok(()) }
    Mode::B => { b(); Ok(()) }
}
```
becomes
```rust
match mode {
    Mode::A => { a(); }
    Mode::B => { b(); }
}
Ok(())
```

This came up from a real cleanup ([uutils/coreutils#13129](https://github.com/uutils/coreutils/pull/13129)) where a function repeated `Ok(())` in every arm.

### Guards (to keep it correct and quiet)

- Every arm body must be a block with a trailing expression; all tails must be `HirEq`.
- The tail must not reference any binding introduced inside the arm (pattern bindings or block `let`s).
- Trivial `()` tails and the all-arms-identical case are skipped.
- A machine-applicable fix is only offered when the `match` is the trailing expression of its enclosing block, so appending the expression after the `match` is always valid. When the value is used elsewhere (e.g. bound in a `let`), the lint stays quiet.

### Lintcheck

Ran filtered to this lint:
- Default crate set (25 crates): 5 new `match` hits (all in `wasmi`), 0 false positives.
- Top 200 most-downloaded crates: 3 new `match` hits (`cc`, `regex-syntax` ×2), 0 false positives.

All inspected hits are genuine and the fixes compile.

### Notes

- This does **not** catch the exact uutils PR above: there one arm's tail was an `if/else` with `Err(..)` in one branch, so the shared `Ok(())` wasn't a direct arm tail — that needs early-return restructuring, out of scope here. It catches the simpler, common sibling of that pattern.
- `branches_sharing_code` is `nursery`, so this is opt-in.

changelog: [`branches_sharing_code`]: also lint `match` expressions whose arms end with the same expression

### CODE REVIEW

File: clippy_lints/src/ifs/branches_sharing_code.rs [hunk 1]
```diff
@@ -100,6 +103,111 @@ pub(super) fn check<'tcx>(
...
+    let first = blocks[0].1;
+    let mut eq = SpanlessEq::new(cx);
+    if !blocks[1..]
+        .iter()
+        .all(|&(_, t)| eq.eq_expr(SyntaxContext::root(), first, t))
+    {
+        return;
+    }
+
+    // Don't bother for a `()` tail: nothing worth moving.
```
Comment (llogiq) [inline clippy_lints/src/ifs/branches_sharing_code.rs:167]: We should put this check above the equality check, it's far cheaper.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17313/304793d2ca1ce788d8f5a68a985f568f1640e528..572a5884c7cf16eba3b721fed00d0d1b1921b511)*

File: clippy_lints/src/ifs/branches_sharing_code.rs [hunk 2]
```diff
@@ -100,6 +103,111 @@ pub(super) fn check<'tcx>(
...
+    if matches!(first.kind, ExprKind::Tup([])) {
+        return;
+    }
+
+    // Require some code before the tail, else hoisting just leaves empty arms.
+    if blocks.iter().all(|(block, _)| block.stmts.is_empty()) {
+        return;
```
Comment (llogiq) [inline clippy_lints/src/ifs/branches_sharing_code.rs:174]: Shouldn't we then suggest removing the `match` wholesale?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17313/304793d2ca1ce788d8f5a68a985f568f1640e528..572a5884c7cf16eba3b721fed00d0d1b1921b511)*

File: tests/ui/branches_sharing_code/shared_match_tail.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            let _g = Guard;
+            kill();
+            Ok(())
+        },
+        _ => {
+            table();
+            Ok(())
+        },
+    }
+}
```
Comment (llogiq) [inline tests/ui/branches_sharing_code/shared_match_tail.rs:115]: Can we also have a test where the arms would all become empty?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17313/304793d2ca1ce788d8f5a68a985f568f1640e528..572a5884c7cf16eba3b721fed00d0d1b1921b511)*

File: clippy_lints/src/ifs/mod.rs
Comment (llogiq) [inline clippy_lints/src/ifs/mod.rs:21]: There should likely be a `Known Problems` section.

* The lint does not trigger if the `match` would have all empty arms
* The lint ignores expanded code

There may be other existing problems with the lint.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17313/304793d2ca1ce788d8f5a68a985f568f1640e528..572a5884c7cf16eba3b721fed00d0d1b1921b511)*

**Review by llogiq — COMMENTED** [review]
I like the extension. The only thing I'd like to see is a few test cases with comments in the match arms (or even before the `=>`).

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17313/409fc810212cea849dc6667b90cd5d6a01c36c32..3f94bf05d2589d7faec12e72c080a32d58885046)*

**Comment by sylvestre** [comment]
@llogiq done! thanks

**Comment by sylvestre** [comment]
@llogiq can we merge it now? :)
Thanks sorry for the pressure

**Review by llogiq — APPROVED** [review]
Yes we can. 👷‍♀️

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17313/e89e38e234e026174b0c87e8c434ac083d764467..4dd09d6ce5b49eb7ea2d55718d742a638adc1f10)*

**Comment by sylvestre** [comment]
Many thanks 🎉
